### DISCUSSION TITLE
FileFilter{}s are case-sensitive

### DISCUSSION THREAD

**Discussion by SolarLune** [ncruces/zenity#8] [Q&A] [answered]
Hello, there!

This is a bit of an interesting finding; it seems that `zenity.FileFilter{}`s are case-sensitive, at least on Linux. I can't, say, find a `.ttf` file if it has `.TTF` as a file extension and the FileFilter is set to specifically `*.ttf` as a pattern. 

It seems like it would be good to make FileFilters case-insensitive for usability. What do you think?

**Comment by ncruces** [comment]
We're just following platform conventions.

On Windows/macOS the file dialogs are case insensitive, but so are their default file systems.
On Linux file systems are case sensitive, and so are globs. The `zenity` command the package wraps on Unix is case sensitive, and to my knowledge there's no option to enable case insensitivity.

So, the only way for this to work would be to pass multiple filters with all possible mixed case combinations (or perhaps just two: lower and uppercase). This is admittedly ugly, but you can already do this yourself.

**Comment by ncruces** [comment]
Actually, looking at the [documentation](https://developer.gnome.org/pygtk/stable/class-gtkfilefilter.html) your pattern could be: `*.[Tt][Tt][Ff]`.

Not sure if I should do this automatically. Would be more discoverable, but then I'd need to go into converting these syntaxes (a bit of which I'm already doing to support macOS).

The way it's currently done is (for the most part): "this supports the patterns allowed by the OS, if you want anything fancy, do a per OS pattern."

But I'll think some more about it.

**Comment by ncruces** [comment, marked as answer]
Latest version adds "support" for character classes to all platforms.

This means you can use the pattern `*.[Tt][Tt][Ff]` portably across all platforms:
- on Linux `*.[Tt][Tt][Ff]` is a case-insensitive match for files with a `TTF` extension;
- on Windows it becomes `*.TTF`, which does the same;
- on macOS it becomes `TTF`, same effect.

I did not add a non-default case-insensitive option to Linux. Making unicode character classes case-insensitive would be error prone, and a source of probable bugs. I'd like to keep those at a minimum.

**Comment by SolarLune** [comment]
Ah, very cool - thanks a lot for implementing this! Sorry for the late reply; I'll see about implementing the above pattern in my code to have case-insensitive filename patterns.

**Comment by ncruces** [comment]
Finally got round to implementing this, hopefully it doesn't introduce too many bugs.
Setting `CaseFold` on a [`FileFilter`](https://pkg.go.dev/github.com/ncruces/zenity#FileFilter) makes it case insensitive on Linux.
